dimod/generators/fcl.py

```diff
@@ -0,0 +1,155 @@
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# ================================================================================================
+from __future__ import absolute_import
+
+import itertools
+import random
+
+from dimod.binary_quadratic_model import BinaryQuadraticModel
+from dimod.decorators import graph_argument
+from dimod.vartypes import SPIN
+
+__all__ = ['frustrated_loop']
+
+
+@graph_argument('graph')
+def frustrated_loop(graph, num_cycles, R=float('inf'), cycle_predicates=tuple(), max_failed_cycles=100):
+    """Generate a frustrated loop problem.
+
+    A (generic) frustrated loop (FL) problem is a sum of Hamiltonians, each generated from a single
+    "good" loop.
+        1. Generate a loop by random walking on the support graph.
+        2. If the cycle is "good" (according to provided predicates), continue, else go to 1.
+        3. Choose one edge of the loop to be anti-ferromagnetic; all other edges are ferromagnetic.
+        4. Add the loop's coupler values to the FL problem.
+    If at any time the magnitude of a coupler in the FL problem exceeds a given precision `R`,
+    remove that coupler from consideration in the loop generation procedure.
+
+    This is a generic generator of FL problems that encompasses both the original FL problem
+    definition from [HJARTL]_ and the limited FL problem definition from [KLH]_
+
+    Args:
+        graph (int/tuple[nodes, edges]/:obj:`~networkx.Graph`):
+            The graph to build the frustrated loops on. Either an integer n, interpreted as a
+            complete graph of size n, or a nodes/edges pair, or a NetworkX graph.
+
+        num_cyles (int):
+            Desired number of frustrated cycles.
+
+        R (int, optional, default=inf):
+            Maximum interaction weight.
+
+        cycle_predicates (tuple[function], optional):
+            An iterable of functions, which should accept a cycle and return a bool.
+
+        max_failed_cycles (int, optional, default=100):
+            Maximum number of failures to find a cycle before terminating.
+
+    .. [HJARTL] Hen, I., J. Job, T. Albash, T.F. Rønnow, M. Troyer, D. Lidar. Probing for quantum
+        speedup in spin glass problems with planted solutions. https://arxiv.org/abs/1502.01663v2
+
+    .. [KLH] King, A.D., T. Lanting, R. Harris. Performance of a quantum annealer on range-limited
+        constraint satisfaction problems. https://arxiv.org/abs/1502.02098
+
+    """
+    nodes, edges = graph
+    assert num_cycles > 0
+    assert R > 0
+    assert max_failed_cycles > 0
+
+    # G = nx.Graph(edges)
+    # J = collections.defaultdict(int)
+    adj = {v: set() for v in nodes}
+    for u, v in edges:
+        if u in adj:
+            adj[u].add(v)
+        else:
+            adj[u] = {v}
+        if v in adj:
+            adj[v].add(u)
+        else:
+            adj[v] = {u}
```

Comment:
For lines 78-85, perhaps instead of if-else's, it might be better to use try-except's. That way, you won't need to search for `u` or `v` in `adj` on each iteration.

i.e.
```
try:
  adj[u].add(v)
except KeyError:
  adj[u] = {v}
```

Reply:
On second thought, this is just a style choice. It won't make a huge difference either way, so feel free to ignore the try-except suggestion.

Reply:
try-catch is faster for the try block and slower for the catch block. In this case I expect to hit the catch block fairly often